whisper/common_test.go

```diff
@@ -0,0 +1,69 @@
+// Contains some common utility functions for testing.
+
+package whisper
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+)
+
+// randomNodeID generates and returns a random P2P discovery node id for the
+// whisper tests.
+func randomNodeID() (id discover.NodeID) {
+	for i := range id {
+		id[i] = byte(rand.Intn(255))
+	}
+	return id
+}
+
+// randomNodeName generates and returns a random P2P node name for the whisper
+// tests.
+func randomNodeName() string {
+	return common.MakeName(fmt.Sprintf("whisper-go-test-%3d", rand.Intn(999)), "1.0")
+}
+
+// whisperCaps returns the node capabilities for running the whisper sub-protocol.
+func whisperCaps() []p2p.Cap {
+	return []p2p.Cap{
+		p2p.Cap{
+			Name:    protocolName,
+			Version: uint(protocolVersion),
+		},
+	}
+}
+
+// bufMsgPipe creates a buffered message pipe between two endpoints.
+func bufMsgPipe() (*p2p.MsgPipeRW, *p2p.MsgPipeRW) {
+	A, midA := p2p.MsgPipe()
+	midB, B := p2p.MsgPipe()
+
+	go copyMsgPipe(midA, midB)
+	go copyMsgPipe(midB, midA)
+
+	return A, B
+}
```

Comment:
This is basically a one-element buffer. Why does it need a buffered pipe? I'm guessing that it's because the handshake message is sent on both sides before receiving. If that's the only reason, it might make sense to change
the protocol instead so it does something like this:

```go
werr := make(chan error, 1)
go func() {
    werr <- p2p.SendItems(ws, statusMsg, protocolVersion)
}()
msg, err := ws.ReadMsg()
if err != nil {
    ...
}
if err := <-werr; err != nil {
    ...
}
```